File: java-code/src/Solns/same_tree_100.java
package Solns;

public class same_tree_100 {
    // Base cases      
    // Both trees are empty, return true
    if(p == null && q == null)
        return true;
    
    // One but not both of the trees are empty (and therefore the two trees are not the same)
    else if(p == null || q == null)
        return false;
    
    // Both trees contain values however the values are not the same 
    else if(p.val != q.val)
        return false;
    
    // Recursively call and if returned values for left and right components of each respective tree are the same, return true
    // If they differ, return false
    return isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
}

File: python-code/Tree_DFS/same_tree_100.py
import TreeNode as T3


def isSameTree(p, q):
    print('p = {}   q = {}'.format(p.val,q.val))
    if not p and not q:
        return True
    elif not p or not q:
        return False

    elif p.val != q.val:
        return False

    return isSameTree(p.left, q.left) and isSameTree(p.right, q.right)



def main():
    print('hi')
    root1 = T3.TreeNode(1)
    root1.left = T3.TreeNode(2)
    # root1.right = T3.TreeNode(3)

    root2 = T3.TreeNode(1)
    root2.left = T3.TreeNode(None)
    root2.right = T3.TreeNode(3)

    print("same tree ? " + str(isSameTree(root1,root2)))

main()
